server/src/main/java/org/opensearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -525,6 +549,78 @@ public static void validateV2TemplateRequest(Metadata metadata, String name, Com
+            );
+        }
+
+        if (isContextAllowed
+            && template.composedOf().stream().anyMatch(componentTemplate -> isSystemTemplate(componentTemplates.get(componentTemplate)))) {
+            throw new InvalidIndexTemplateException(
+                name,
+                "index template [" + name + "] specifies a component templates which can only be used in context."
+            );
+        }
+
+        if (template.context() != null && findContextTemplate(metadata, template.context()) == null) {
+            throw new InvalidIndexTemplateException(
+                name,
+                "index template [" + name + "] specifies a context which is not loaded on the cluster."
+            );
+        }
+    }
+
+    private void validateComponentTemplateRequest(ComponentTemplate componentTemplate) {
+        if (isSystemTemplate(componentTemplate)
+            && !ClusterStateSystemTemplateLoader.TEMPLATE_LOADER_IDENTIFIER.equals(
+                threadPool.getThreadContext().getTransient(ACTION_ORIGIN_TRANSIENT_NAME)
+            )) {
+            throw new IllegalArgumentException("A system template can only be created with a repository");
+        }
+    }
+
+    private static String findContextTemplate(Metadata metadata, Context context) {
+        if (context == null) {
+            return null;
+        }
+        long maxVersionSeen = -1L;
+        final boolean searchSpecificVersion = !Context.LATEST_VERSION.equals(context.version());
+
+        String systemTemplateToUse = null;
+        for (Map.Entry<String, ComponentTemplate> entry : metadata.componentTemplates().entrySet()) {
+            if (isSystemTemplate(entry.getValue())) {
+                SystemTemplateMetadata templateMetadata = SystemTemplateMetadata.fromComponentTemplate(entry.getKey());
+                if (templateMetadata.name().equals(context.name())) {
+                    if (!searchSpecificVersion && maxVersionSeen < templateMetadata.version()) {
+                        maxVersionSeen = templateMetadata.version();
+                        systemTemplateToUse = entry.getKey();
+                    } else if (searchSpecificVersion && templateMetadata.version() == Long.parseLong(context.version())) {
+                        systemTemplateToUse = entry.getKey();
+                    }
+                }
+            }
+        }
+        return systemTemplateToUse;
+    }
+
+    private static boolean isSystemTemplate(ComponentTemplate componentTemplate) {
+        return Optional.ofNullable(componentTemplate)
+            .map(ComponentTemplate::metadata)
+            .map(md -> md.get(ClusterStateSystemTemplateLoader.TEMPLATE_TYPE_KEY))
+            .filter(ob -> SystemTemplateMetadata.COMPONENT_TEMPLATE_TYPE.equals(ob.toString()))
+            .isPresent();
+    }
+
+    private static boolean hasOverlap(Settings.Builder userCollection, Settings baseSettings) {
```

Comment:
This is to address conflicts. 

Based on the proposal https://github.com/opensearch-project/OpenSearch/issues/14649, 

```
If there is a conflict between what is declared through an index template and through ABC templates, we can either fail during simulation or handle gracefully.
```

I've followed the path to fail in case of overlap. We can discuss this if needed

Reply:
Ad discussed around the use-cases for dynamic settings, I'll honor the precedence level.